pymc3/distributions/distribution.py

```diff
@@ -141,40 +245,116 @@ def __new__(cls, name, *args, **kwargs):
                 "for a standalone distribution."
             )
 
-        rng = kwargs.pop("rng", None)
-
-        if rng is None:
-            rng = model.default_rng
-
         if not isinstance(name, string_types):
             raise TypeError(f"Name needs to be a string but got: {name}")
 
-        data = kwargs.pop("observed", None)
-
-        total_size = kwargs.pop("total_size", None)
-
-        dims = kwargs.pop("dims", None)
-
-        if "shape" in kwargs:
-            raise DeprecationWarning("The `shape` keyword is deprecated; use `size`.")
+        if rng is None:
+            rng = model.default_rng
 
-        transform = kwargs.pop("transform", UNSET)
+        # Create the RV without specifying testval, because the testval may
+        # have a shape that only matches after replicating with a size implied
+        # by dims (see below).
+        rv_out = cls.dist(*args, rng=rng, testval=None, **kwargs)
+
+        # `dims` are only available with this API, because `.dist()` can be used
+        # without a modelcontext and dims are not yet tracked at the Aesara level.
+        if dims is not None:
+            # Infer size from dims (and coords)
+            size = tuple(len(model.coords[dname]) for dname in dims if dname != Ellipsis)
```

Comment:
I think this line can be moved inside the `if` in order to allow this usecase, I personally will mostly use dims without ellipsis (I already use `dims` in deterministics only for explicitness sake and to get everything converted nicely to ArviZ without extrra args) and 1) would love doing that 2) I think it's powerful and conceptually clear feature.

Reply:
Not sure I got your point completely.
> why should I provide coords at all if like with obs_id it is an implied dim and I'm already fine with coord values being integers?
> why should I provide coords at model creation if I plan to modify those coords every time I fit the model? (i.e. to explore different combinations of covariates)

You shouldn't need to do any of the above. But in `Distribution.__new__` the length of the implied dimensions is not necassarily fixed (right?), so that's not the place where to auto-define the integer coords.
That's why my question is more like: "_Can we auto-create coords missing for individual dimensions (where the entry `is None`) at the time of `InferenceData` conversion?_"

With `pm.Data` I would propose to add a `coords` dict kwarg to `set_value` so swapping data and the corresponding coords is the same command.

Reply:
> You shouldn't need to do any of the above

The issue is that I currently have to, because I am using dim names for implied dimensions and the current code will try to do `len(model.coords[dname])` so I need to provide those coords in `pm.Model`. 

> But in Distribution.__new__ the length of the implied dimensions is not necassarily fixed (right?)

I think so

> so that's not the place where to auto-define the integer coords.
That's why my question is more like: "Can we auto-create coords missing for individual dimensions (where the entry is None) at the time of InferenceData conversion?"

This is more related to my comment below, we shouldn't auto-create coords at all, ever. ArviZ will automatically handle that. In fact, we shouldn't even need to auto create dim names either, both these tasks are basically duplicating ArviZ code in a place where it should not be so we need to resort to tricks like using `None`. As I commented below, the way to indicate that some coordinate values should be auto-created is by the dim name not being present in the coords dict.

> With pm.Data I would propose to add a coords dict kwarg to set_value so swapping data and the corresponding coords is the same command.

This sounds great

Reply:
> [...] and the current code will try to do len(model.coords[dname]) so I need to provide those coords in pm.Model
Oh now I get it!
Okay, here's the challenge:

We don't know how many dimensions were _not_ passed unless `Ellipsis` is used.
Without `Ellipsis` we must assume that `dims` corresponds to all dimensions.

To figure out `size`, at least the first `len(dims) - ndim_implied` `dims` _must_ be known to the model.
Unless `size` can be symbolic. I'm not sure about that at the moment.

Reply:
Yes, `size` can definitely be symbolic.

Reply:
> We don't know how many dimensions were _not_ passed unless `Ellipsis` is used.
> Without `Ellipsis` we must assume that `dims` corresponds to all dimensions.

Nothing should be made to depend on this`Ellipsis` feature.  If that's the case, then there's a serious design issue somewhere.

Reply:
This is just saying that if there are no Ellipses in the tuple, there is not currently a good way to tell the user if they mixed up the implied dims.

With shape, a mixup would raise a somewhat informative AssertionError in the first iteration. Mixed-up dims will just crash the conversion to InferenceData.

It's really not a problem from our perspective.